### PR TITLE
Opaque relay op codegen

### PR DESCRIPTION
**Pull request by hypercubestart**
TODO:
- [x] implement codegen for relay ops
- [x] implement add broadcasting
- [x] add tests
  - [x] RelayGlobalAvgPool2D
  - [x] RelayBatchFlatten
  - [x] RelayBiasAdd
- [x] implement assertion to check `result` with `out` array (need to add functions on NDarray)
  - [x] batchnorm
  - [x] softmax
  - [x] relu
  - [x] maxpool2d

codegen for #65

### CODE REVIEW

**Comment by hypercubestart** [comment]
@gussmith23 build is failing because CI/CD machine doesn't have "math.h"

**Comment by gussmith23** [comment]
Wait until #73 merged

**Comment by gussmith23** [comment]
Interesting about math.h...odd that it wouldn't have it. Will debug later!

File: src/language/from_relay/mod.rs
Comment (gussmith23) [inline src/language/from_relay/mod.rs:709]:
```suggestion
                    // Should be ___C or _C___ (e.g. NHWC, NCHW)
```

File: src/language/from_relay/run_relay.py
Comment (gussmith23) [inline src/language/from_relay/run_relay.py:36]:
```suggestion
```

**Review by gussmith23 — CHANGES_REQUESTED** [review]
Looking good, let's iterate!

File: src/codegen.rs
Comment (hypercubestart) [reply src/codegen.rs:760]: yep, to not hardcode shape, the maxpool c function needs to be extended for other shapes

File: src/codegen.rs
Comment (hypercubestart) [reply src/codegen.rs:824]: dont quite follow here. global avg pool 2d averages across height/width layers, so an input dim= 4 NHWC tensor becomes a dim = 2 NC tensor

Comment (hypercubestart) [reply src/codegen.rs:824]: i dont think quite follow here. The input tensor is NHWC, and the output tensor is expected to be NC. Is there another way to getting the shape?

File: src/codegen.rs
Comment (hypercubestart) [reply src/codegen.rs:3470]: not much of a better solution maybe, but I just added explicit casts

File: src/codegen.rs
Comment (hypercubestart) [reply src/codegen.rs:3606]: i think this should be included for debugging sake...

File: src/codegen.rs
Comment (hypercubestart) [reply src/codegen.rs:4149]: works!

File: src/codegen.rs
Comment (hypercubestart) [reply src/codegen.rs:3569]: i think this should be included for debugging sake...

File: c-files/relay-op-implementations.c
Comment (gussmith23) [inline c-files/relay-op-implementations.c:25]:
```suggestion
  // TODO: (1/sqrt(var+epsilon) * gamma) can be pre-computed
```

File: src/codegen.rs
Comment (gussmith23) [inline src/codegen.rs:569]:
```suggestion
```
I assume we don't need this anymore? Can you check for other leftover print statements?

File: src/codegen.rs
Comment (gussmith23) [inline src/codegen.rs:3004]:
```suggestion
```
Seems like it's out of date?

File: src/codegen.rs
Comment (gussmith23) [inline src/codegen.rs:3141]:
```suggestion
```

I think this one is also out of date?

File: src/codegen.rs
Comment (gussmith23) [inline src/codegen.rs:3342]: What does this mean? Is it still relevant?

File: src/codegen.rs
Comment (gussmith23) [inline src/codegen.rs:3332]:
```suggestion
```
Remove this and other printlns

File: src/codegen.rs
Comment (gussmith23) [inline src/codegen.rs:899]: Are these both still relevant?

File: src/codegen.rs
Comment (gussmith23) [inline src/codegen.rs:3511]: Another one -- is this still relevant?

File: src/codegen.rs
Comment (gussmith23) [reply src/codegen.rs:3563]: Sounds good!

**Review by gussmith23 — COMMENTED** [review]
Nearly there, just a few questions!

File: src/codegen.rs
Comment (gussmith23) [inline src/codegen.rs:630]: Hm. The input shape here is `(1, 1000)`, right? As long as the first dimension, batch, is 1, then our softmax kernel works. Can you add an assert here to assert that batch is 1?

Comment (gussmith23) [reply src/codegen.rs:630]: That is, if we have batch > 1, the kernel breaks (because it will softmax over EVERYTHING rather than splitting the kernel up), but as long as batch == 1, then we're good.

**Review by gussmith23 — COMMENTED** [review]
Can you run the formatter, too? Looks like formatting got out of whack.

Plus one last comment, then we're good!

File: src/codegen.rs
Comment (hypercubestart) [reply src/codegen.rs:630]: is the shape guaranteed to be two dimensions though?

Comment (gussmith23) [reply src/codegen.rs:630]: Yeah, that should be a fine assumption for now. All the CNNs we're currently working with use softmax as their very last layer, and the input/output shape is `(batch, num_classes)`. You could also assert that the number of dimensions is 2, if you want to be extra safe. That should be the case for all of our current networks.

Comment (gussmith23) [reply src/codegen.rs:630]: FWIW, my general strategy is that it's okay if we put in asserts that OVER-constrain. Then, at the very worst, you get a false-positive assertion failure, and you can just refine the condition in the assertion. In your current code, I don't think there's anything to stop us from generating C code that just straight up breaks, if the batch size is greater than 1 in the softmax, or if the `axis` param from relay is something out of the norm. Then, the bug would come in the form of an error from the C code and would be harder to track down!

**Review by gussmith23 — APPROVED** [review]

**Pull request merged**
